src/DNS_Mgr.cc

```diff
@@ -40,418 +42,613 @@
-	int map_type;
-	bool no_mapping; // when initializing from a file, immediately hit EOF
-	bool init_failed;
-	bool failed;
-	};
+		unsigned char* query = NULL;
+		int len = 0;
+		int status = ares_create_query(query_host.c_str(), C_IN, request_type,
+		                               DNS_Request::request_id, 1, &query, &len, 0);
 
-void DNS_Mgr_mapping_delete_func(void* v)
-	{
-	delete (DNS_Mapping*)v;
-	}
+		if ( status != ARES_SUCCESS )
+			return;
 
-static TableValPtr empty_addr_set()
-	{
-	auto addr_t = base_type(TYPE_ADDR);
-	auto set_index = make_intrusive<TypeList>(addr_t);
-	set_index->Append(std::move(addr_t));
-	auto s = make_intrusive<SetType>(std::move(set_index), nullptr);
-	return make_intrusive<TableVal>(std::move(s));
+		// Store this so it can be destroyed when the request is destroyed.
+		this->query = query;
+		ares_send(channel, query, len, query_cb, req_data);
+		}
 	}
 
-DNS_Mapping::DNS_Mapping(const char* host, struct hostent* h, uint32_t ttl)
+void DNS_Request::ProcessAsyncResult(bool timed_out, DNS_Mgr* mgr)
 	{
-	Init(h);
-	req_host = util::copy_string(host);
-	req_ttl = ttl;
+	if ( ! async )
+		return;
 
-	if ( names && ! names[0] )
-		names[0] = util::copy_string(host);
+	if ( request_type == T_A )
+		mgr->CheckAsyncHostRequest(host, timed_out);
+	else if ( request_type == T_PTR )
+		mgr->CheckAsyncAddrRequest(addr, timed_out);
+	else
+		mgr->CheckAsyncOtherRequest(host, timed_out, request_type);
 	}
 
-DNS_Mapping::DNS_Mapping(const IPAddr& addr, struct hostent* h, uint32_t ttl)
+/**
+ * Retrieves the TTL value from the first RR in the response.
+ *
+ * This code is adapted from an internal c-ares method called * ares__parse_into_addrinfo,
+ * which is used for ares_getaddrinfo callbacks. It's also the only method that properly
+ * parses out TTL data currently. This skips over the question and the first bit of the
+ * response to get to the first RR, and then returns the TTL from that RR. We only use the
+ * first RR because it's a good approximation for now, at least until the work in c-ares
+ * lands to add TTL support to the other RR-parsing methods.
+ *
+ * @param abuf The buffer containing the entire response from the server.
+ * @param alen The length of the buffer
+ * @param ttl An out param for returning the TTL value.
+ * @return A status code from c-ares. This will be ARES_SUCCESS on success, or some other
+ * code on failure.
+ */
+static int get_ttl(unsigned char* abuf, int alen, int* ttl)
 	{
-	Init(h);
-	req_addr = addr;
-	req_host = nullptr;
-	req_ttl = ttl;
-	}
+	int status;
+	long len;
+	char* hostname = NULL;
 
-DNS_Mapping::DNS_Mapping(FILE* f)
-	{
-	Clear();
-	init_failed = true;
+	*ttl = DNS_TIMEOUT;
 
-	req_host = nullptr;
-	req_ttl = 0;
-	creation_time = 0;
+	unsigned char* aptr = abuf + HFIXEDSZ;
+	status = ares_expand_name(aptr, abuf, alen, &hostname, &len);
+	if ( status != ARES_SUCCESS )
+		return status;
+	if ( aptr + len + QFIXEDSZ > abuf + alen )
+		return ARES_EBADRESP;
```

Comment:
Can this leak hostname (i.e., missing `ares_free_string()` call).

I see this below too and am wondering if we should add RAII wrappers for c-ares types to avoid leaks. Seems that could happen easily and have quite an impact in a bad case.

Reply:
I can do something like this:

```
using ares_str = std::unique_ptr<char, std::function<void(char*)>>;
ares_str hostname{nullptr, [](char* s){ares_free_string(s);}};

```

The only problem here is that I can't pass `&hostname.get()` to a method since it won't let me take the address of the rvalue there, but that can be worked around by just taking the pointer and storing it in a temporary lvalue.

Reply:
> This isn't terribly safe since you're changing the pointer out from underneath unique_ptr and it likely won't free correctly.

Maybe something like this https://github.com/soasis/out_ptr? It was apparently added into c++23 recently: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1132r8.html.

Reply:
I like this library and I could see bringing it in as well as using it in the zeek-agent-v2 work. I defined the following:

```
struct ares_deleter
	{
	void operator()(char* s) const { ares_free_string(s); }
	void operator()(unsigned char* s) const { ares_free_string(s); }
	void operator()(ares_addrinfo* s) const { ares_freeaddrinfo(s); }
	void operator()(struct hostent* h) const { ares_free_hostent(h); }
	void operator()(struct ares_txt_reply* h) const { ares_free_data(h); }
	};
```

And now I can do things like:

```
std::unique_ptr<struct hostent, ares_deleter> he;
status = ares_parse_ptr_reply(buf, len, &addr, sizeof(addr), AF_INET, out_ptr<struct hostent*>(he));
```

and 

```
std::unique_ptr<char, ares_deleter> hostname;
status = ares_expand_name(aptr, abuf, alen, out_ptr<char*>(hostname), &len);
```

Reply:
This is neat, I like it!

Reply:
I think that's everything in your review then. I'm going to pass this latest version off to @dopheide-esnet and @initconf for some more testing, and then move this PR out of draft status.